File: abc017/d.py
N, K = map(int, input().split())
F = [int(input()) for i in range(N)]
s = set()
A = [0] * (N + 1)
A[-1] = 1
cu = 1
mod = 10 ** 9 + 7

left = 0
for i, f in enumerate(F):
    while f in s:
        s.remove(F[left])
        cu -= A[left - 1]
        cu %= mod
        left += 1
    s.add(f)
    A[i] = cu
    cu += cu
    cu %= mod
print(A[-2] % mod)

File: abc017/d.rb
MOD = 1_000_000_007

n, m = gets.split.map(&:to_i)
fs = Array.new(n) { gets.to_i - 1 } # 0-origin

taken = Array.new(m)
lefts = Array.new(n + 1, 0)

# 尺取り法
left = 0
n.times do |right|
  while left < right && taken[fs[right]]
    taken[fs[left]] = false
    left += 1
  end
  taken[fs[right]] = true
  lefts[right + 1] = left
end

# dp[0] = 1
# dp[i] = dp[j...i].sum
# [j,i) is unique
dp = Array.new(n + 1)
dp[0] = 1
# cum sum
# cs[i+1] = dp[0..i].sum
cs = Array.new(n + 2)
cs[0] = 0
cs[1] = 1

(1..n).each do |i|
  dp[i] = (cs[i] - cs[lefts[i]]) % MOD
  cs[i + 1] = (cs[i] + dp[i]) % MOD
end

# p dp.zip(lefts, 0.step).map { |(c, l, r)| [c, l...r] }
p dp[-1]
